---  
# Question 1

Contributions from Question 1:
- Jason Li: Wrote and tested code

In [39]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.interpolate import lagrange
from scipy.optimize import curve_fit

## 1a) Approximate f(1.4) for 4 different polynomials

In [53]:
x_i = [1,1.25,1.6,1.8] 
def f(x): return np.sin(np.pi * x)
x = 1.4
y_i = [f(x_i[0]),  f(x_i[1]), f(x_i[2]), f(x_i[3])]
p_12 = lagrange(x_i[1:3],y_i[1:3])
p_012 = lagrange(x_i[0:3],y_i[0:3])
p_123 = lagrange(x_i[1:4],y_i[1:4])
p_0123 = lagrange(x_i[0:4],y_i[0:4]) 
y_12 = p_12(x)
y_012 = p_012(x)
y_123 = p_123(x)
y_0123 = p_0123(x)
print("P_12(1.4): ", y_12) 
print("P_012(1.4): ", y_012)
print("P_123(1.4): ", y_123)
print("P_0123(1.4): ", y_0123)
print("True value: ", f(x))


P_12(1.4):  -0.8116566676616641
P_012(1.4):  -0.9182280617406038
P_123(1.4):  -0.948748789289704
P_0123(1.4):  -0.9334884255151685
True value:  -0.9510565162951535


$P_{1,2}(1.4)$:  -0.8116566676616641

$P_{0,1,2}(1.4)$:  -0.9182280617406038

$P_{1,2,3}(1.4)$:  -0.948748789289704

$P_{0,1,2,3}(1.4)$:  -0.9334884255151685

True value:  -0.9510565162951535


## 1b) Compute the absolute error for each polynomial

In [49]:
diff_12 = np.abs(y_12 - f(x))
diff_012 = np.abs(y_012 - f(x))
diff_123 = np.abs(y_123 - f(x))
diff_0123 = np.abs(y_0123 - f(x))
print("Absolute error of P_12(1.4): ", diff_12) 
print("Absolute error of P_012(1.4): ", diff_012)
print("Absolute error of P_123(1.4): ", diff_123)
print("Absolute error of P_0123(1.4): ", diff_0123)

Absolute error of P_12(1.4):  0.13939984863348942
Absolute error of P_012(1.4):  0.032828454554549746
Absolute error of P_123(1.4):  0.002307727005449567
Absolute error of P_0123(1.4):  0.017568090779985


Absolute error of $P_{1,2}$(1.4):  0.13939984863348942

Absolute error of $P_{0,1,2}$(1.4):  0.032828454554549746

Absolute error of $P_{1,2,3}$(1.4):  0.002307727005449567

Absolute error of $P_{0,1,2,3}$(1.4):  0.017568090779985

Based on these values, we can see that the interpolating quadratic ($P_{1,2,3}$(x)) is best since it has the lowest absolute error

## 1c) Error bounds for each polynomial

In [50]:
def f2(x): return -np.sin(np.pi * x) * (np.pi ** 2) # second derivative
def f3(x): return -np.cos(np.pi * x) * (np.pi ** 3) # third derivative
def f4(x): return np.sin(np.pi * x) * (np.pi ** 4) # fourth derivative
interval12 = [x_i[1],x_i[2]]
interval012 = [x_i[0],x_i[2]]
interval123 = [x_i[1],x_i[3]]
interval0123 = [x_i[0],x_i[3]]
maxf2_12 = np.max(abs(f2(np.linspace(interval12[0], interval12[1], 10000)))) # max of second derivative on interval 1.25,1.6
    # true max for above is at abs(f2(1.5)) when derived by hand
maxf3_012 = np.max(abs(f3(np.linspace(interval012[0], interval012[1], 10000)))) # max of third derivative on interval 1, 1.6
    # true max for above is at abs(f3(1)) when derived by hand
maxf3_123 = np.max(abs(f3(np.linspace(interval123[0], interval123[1], 10000)))) # max of third derivative on interval 1.25, 1.8
    # true max for above is at abs(f3(1.8)) when derived by hand
maxf4_0123 = np.max(abs(f4(np.linspace(interval012[0], interval012[1], 10000)))) # max of fourth derivative on interval 1, 1.8
    # true max for above is at abs(f4(1.5)) when derived by hand

# compute error terms
eb12 = abs(maxf2_12 / (2 * 1) * (x-x_i[1]) * (x-x_i[2]))
eb012 = abs(maxf3_012 / (3 * 2 * 1) * (x-x_i[0]) * (x-x_i[1]) * (x-x_i[2]))
eb123 = abs(maxf3_123 / (3 * 2 * 1) * (x-x_i[1]) * (x-x_i[2]) * (x-x_i[3]))
eb0123 = abs(maxf4_0123 / (4 * 3 * 2 * 1) * (x-x_i[0]) * (x-x_i[1]) * (x-x_i[2]) * (x-x_i[3]))
print("Error bound for P_12(1.4): ", eb12) 
print("Error bound for P_012(1.4): ", eb012)
print("Error bound for P_123(1.4): ", eb123)
print("Error bound for P_0123(1.4): ", eb0123)

Error bound for P_12(1.4):  0.14804406599807254
Error bound for P_012(1.4):  0.06201255336059963
Error bound for P_123(1.4):  0.05016920953330838
Error bound for P_0123(1.4):  0.01948181812025816


Error bound for $P_{1,2}$(1.4):  0.14804406599807254

Error bound for $P_{0,1,2}$(1.4):  0.06201255336059963

Error bound for $P_{1,2,3}$(1.4):  0.05016920953330838

Error bound for $P_{0,1,2,3}$(1.4):  0.01948181812025816

Based on these values, we can see that the interpolating cubic ($P_{0,1,2,3}$(x)) has the best (lowest) error bound. These error bounds make sense because all the errors in part (b) are less than their corresponding error bounds here. We can see that this result of which polynomial is best is different from part (b).